### PR TITLE
Ben sprint 4

### PR DESCRIPTION
**Pull request by bjhobson12**
This pr allows hands to track lnk using bfs. It doesn't seem to have performance issues and is updated at every tile so the hands can constantly follow him. When a hand has caught link, it will animate out in a relatively cool spiral. Then the game should be reset to the main dungeon... and I really mean reset... there are enough variables changed in handenemy it might be safer to reset. We could try just changing the currentroom, but I have not tested that.

### CODE REVIEW

File: Sprintfinity3902/Dungeon/RoomLoader.cs [hunk 1]
```diff
@@ -47,13 +47,15 @@ public class RoomLoader : IRoomLoader
...
         
 
         // Have this input a filename and then load the room.
-        public RoomLoader(IRoom room)
+        public RoomLoader(IPlayer player)
         {
             // Really think there is a better way to list these files, just a demo for the time being though.
-            Initialize(room);
+            link = player;
         }
```
Comment (kylegessel) [inline Sprintfinity3902/Dungeon/RoomLoader.cs:59]: This looks like a big change to the RoomLoader that I'm not quite sure I understand.

Comment (kylegessel) [reply Sprintfinity3902/Dungeon/RoomLoader.cs:59]: I get the the hand needs to know link's location, but does the Initialize still occur?

File: Sprintfinity3902/Entities/Enemies_NPCs/HandEnemy.cs [hunk 2]
```diff
@@ -9,7 +12,7 @@ namespace Sprintfinity3902.Entities
     public class HandEnemy : AbstractEntity, IEnemy
     {
 
-        private static float F_DOT_TWO = .2f;
+        private static float F_DOT_THREE = .4f;
```
Comment (kylegessel) [inline Sprintfinity3902/Entities/Enemies_NPCs/HandEnemy.cs:15]: Name doesn't match reality anymore. I think we should change our Magic number variables to descriptive names (in this case, something like ENTITY_SPEED) so we don't have to update both the number and the variable name.

File: Sprintfinity3902/Entities/Link/DamagedLink.cs [hunk 3]
```diff
@@ -17,57 +17,60 @@ public class DamagedLink : ILink
...
-            {
-                return _sprite;
+            get {
+                return decoratedLink.Sprite;
             }
-            set
-            {
-                _sprite = value;
+            set {
+                decoratedLink.Sprite = value;
```
Comment (kylegessel) [inline Sprintfinity3902/Entities/Link/DamagedLink.cs:27]: What do these changes give us?

File: Sprintfinity3902/Interfaces/IEntity.cs [hunk 4]
```diff
@@ -21,6 +21,10 @@ float Y
         {
             get; set;
         }
+        bool STATIC 
+        { 
+            get; set; 
+        }
```
Comment (kylegessel) [inline Sprintfinity3902/Interfaces/IEntity.cs:27]: If this isn't a constant, I don't think it should be capitalized.

File: Sprintfinity3902/States/ItemPickup/BlueRupeePickup.cs [hunk 5]
```diff
@@ -14,7 +14,6 @@ public bool Pickup(IPlayer Link)
         {
             Link.itemcount[IItem.ITEMS.RUPEE]+=BLUE_RUPEE_VALUE;
 
-            Link.itemPickedUp = true;
```
Comment (kylegessel) [inline Sprintfinity3902/States/ItemPickup/BlueRupeePickup.cs:17]: This removal in item classes broke the HUD

Comment (kylegessel) [reply Sprintfinity3902/States/ItemPickup/BlueRupeePickup.cs:17]: As it stands, the boolean is how we tell the HUD something has happened. I don't think its ideal, but it is still required in the current code base.

File: Sprintfinity3902/Interfaces/IPlayer.cs [hunk 6]
```diff
@@ -25,20 +25,18 @@ enum SelectableWeapons
...
-
+        void SetState(IPlayerState state);
 
         /*These methods are only used when link is not invinsible (damaged state)*/
         void MoveLink();
         void StopMoving();
         void TakeDamage();
         void BounceOfEnemy(ICollision.CollisionSide Side);
         void Initialize();
+        void TogglePlayerVisible();
```
Comment (kylegessel) [inline Sprintfinity3902/Interfaces/IPlayer.cs:40]: Is this for the updated lose state?

File: Sprintfinity3902/States/ItemPickup/BlueRupeePickup.cs (hunk 5 above)
Comment (kylegessel) [reply Sprintfinity3902/States/ItemPickup/BlueRupeePickup.cs:17]: I think UpdateSelf command is supposed to replace this, but it doesn't look complete.

File: Sprintfinity3902/Entities/Enemies_NPCs/BlueBatEnemy.cs [hunk 7]
```diff
@@ -37,6 +28,8 @@ public BlueBatEnemy(Vector2 pos)
             count = 0;
             speed = .5f;
             SetStepSize(speed);
+            Collidable = false;
+            Collidable = false;
```
Comment (kylegessel) [inline Sprintfinity3902/Entities/Enemies_NPCs/BlueBatEnemy.cs:32]: Looks like an accident here.

**Review by kylegessel — CHANGES_REQUESTED** [review]
Brad and I both had I instant reaction to seeing how the new hand enemies. Watching the hands path to link was great and then the animation after you get caught was just the cherry on top. This is so damn cool.

I made alot of comments reading through this code.

I know this code presently breaks the HUD, which is a major concern. I see that you are making methods that link can call to update the commands, but those don't seem implemented as of now. 

The reason we held off on commands like that is that we decided HUD should see link information but maybe we could avoid Link needing to know anything about HUD. By the nature of the HUD it HAS to know about link anyway (to check his health etc) so creating booleans for the HUD to check felt like an ok solution. I think we should have a discussion about this.

**Comment by bjhobson12** [comment]
This pr is out of date... will re-request when ready. Thank you all for the comments.

**Pull request closed without merging**